### PR TITLE
SDK Updates

### PR DESCRIPTION
**Pull request by harman28**
Refer: [SDK Updates](https://github.com/razorpay/api/issues/2313)

 - Customers
 - - Support to fetch and edit customers
 - Refunds
 - - Support to fetch refunds independent of payment
 - - Support to create refund without fetching payment
 - Payments
 - - Refactored payment refund fetch methods
 - Cards
 - - Support to fetch cards
 - Support to add customer headers
 - Utility function to validate signature received from payment callback

### CODE REVIEW

File: lib/razorpay/card.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'razorpay/request'
+require 'razorpay/entity'
+
+module Razorpay
+  # Card API allows you to fetch cards
```
Comment (captn3m0) [inline lib/razorpay/card.rb]: Put a link to cards api docs here?

File: lib/razorpay/utility.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    private
+
+    def self.validate_signature(signature, *fields)
+      secret = Razorpay.auth[:password]
+
+      data = fields.join('|')
+
+      expected_signature = OpenSSL::HMAC.hexdigest('SHA256', secret, data)
+
+      return expected_signature === signature
```
Comment (captn3m0) [inline lib/razorpay/utility.rb]: We need a constant time comparison here. Ruby doesn't have one inbuilt [yet](https://bugs.ruby-lang.org/issues/10098), so we can copy the one that [Rack has](http://www.rubydoc.info/github/rack/rack/Rack/Utils.secure_compare)

File: README.md [hunk 3]
```diff
@@ -46,6 +51,13 @@ You can handle refunds using the following constructs:
 
 ```rb
 Razorpay::Payment.fetch("payment_id").refund({amount:500})
+refunds = Razorpay::Payment.fetch("payment_id").refunds
+```
+
+Refunds can also be handled without fetching payments:
+```rb
+refund = Razorpay::Refund.create(payment_id:"payment_id")
```
Comment (captn3m0) [inline README.md]: 1. stick to new style hashes wherever possible?
2. space after colon

File: lib/razorpay/payment.rb [hunk 4]
```diff
@@ -24,8 +24,7 @@ def refund(options = {})
     end
 
     def refunds
-      # This needs to be a string, not a symbol
-      Refund.new('payment_id' => id)
+      self.class.request.request :get, "payments/#{id}/refunds"
```
Comment (captn3m0) [inline lib/razorpay/payment.rb]: Can we add a `get` method in Request to make this shorter?

File: lib/razorpay/utility.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    def self.validate_payment_signature(order_id, payment_id, signature)
+      return validate_signature(signature, order_id, payment_id)
+    end
+
+    private
+
+    def self.validate_signature(signature, *fields)
+      secret = Razorpay.auth[:password]
+
+      data = fields.join('|')
```
Comment (captn3m0) [inline lib/razorpay/utility.rb]: brackets not needed

**Comment by captn3m0** [comment]
Ran rubocop:

```
 rubocop 
Inspecting 27 files
...............C...........

Offenses:

lib/razorpay/entity.rb:15:5: C: When using method_missing, define respond_to_missing? and fall back on super.
    def method_missing(name) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^

27 files inspected, 1 offense detected
```

File: README.md (hunk 3 above)
Comment (harman28) [inline README.md]: Isn't this the new style? Is `:symbol => 'value'` preferred over `symbol: 'value'`?

Comment (captn3m0) [inline README.md]: Yes, this is the new style, and I'd prefer `symbol: "Value"`. The old style is in a few other places, and we can change it elsewhere.

File: lib/razorpay/utility.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+require 'razorpay'
+require 'openssl'
+
+# Helper functions are defined here
+module Razorpay
+  class Utility
+    def self.validate_payment_signature(order_id, payment_id, signature)
```
Comment (harman28) [inline lib/razorpay/utility.rb]: We're changing this back to accept an `attributes` hash as the only param, to avoid exactly this case.

Comment (captn3m0) [inline lib/razorpay/utility.rb]: But you are still accepting order_id, no? My point is for the doc to be clear that the order_id passed to this method should not be tainted

File: README.md (hunk 3 above)
Comment (captn3m0) [inline README.md]: @harman28  we should add a doc about the signature helper as well?

Comment (harman28) [inline README.md]: You're right. Adding a section to the README.

**Pull request merged**